prow/kube/prowjob_collector.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kube
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+type prowJobClient interface {
+	List(opts metav1.ListOptions) (*prowapi.ProwJobList, error)
+}
+
+//https://godoc.org/github.com/prometheus/client_golang/prometheus#Collector
+type prowJobCollector struct {
+	name     string
+	client   prowJobClient
+	selector string
+}
+
+func (pjc prowJobCollector) Describe(ch chan<- *prometheus.Desc) {
+	//prometheus.DescribeByCollect(pjc, ch) //Normally, we would need this line.
+	//https://godoc.org/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics
+	//This is a take-our-own-risk action and also a compromise for implementing a metric with
+	//both dynamic keys and dynamic values in the label set.
+	//The formed metrics scraping data have to be tested locally with prometheus server.
+	//Another option is taken for kube_pod_labels: https://github.com/kubernetes/kube-state-metrics/blob/master/docs/pod-metrics.md
+	//If we follow the implementation there, we would need to import pkgs from "k8s.io/kube-state-metrics".
+}
+
+func (pjc prowJobCollector) Collect(ch chan<- prometheus.Metric) {
+	logrus.Debug("ProwJobCollector collecting ...")
+	prowJobs, err := pjc.client.List(metav1.ListOptions{LabelSelector: pjc.selector})
+	if err != nil {
+		logrus.WithError(err).Error("Failed to list prow jobs with selector '%s'", pjc.selector)
+	}
+	for _, pj := range prowJobs.Items {
+		labels := []string{"prow_job_name", "prow_job_namespace"}
+		labelValues := []string{pj.Name, pj.Namespace}
+		for k, v := range pj.Labels {
+			labels = append(labels, k)
+			labelValues = append(labelValues, v)
+		}
+		desc := prometheus.NewDesc(
+			"prow_job_labels",
```

Comment:
@stevekuznetsov 
We want to add annotations as well right?

Would annotations be handled as another metric `prow_job_annotations`?

Reply:
Yes, that sounds good.

Reply:
This PR becomes bigger and bigger.
I will add `prow_job_annotations` as a separate PR if that is OK.

Reply:
This PR becomes bigger and bigger.
I will add `prow_job_annotations` as a separate PR if that is OK.